### PR TITLE
Speed up BundleFuse

### LINKED ISSUE #757: Make cl mount faster
Currently the REST calls dominate the processing times for every read request. Try to find a way to buffer.

### PR DESCRIPTION
**Pull request by fabeschan**
Prefetch byte ranges.

Fixes #757

### CODE REVIEW

**Comment by fabeschan** [comment]
I'll do a clean up once I get an approval for the overall approach

File: codalab/lib/bundle_fuse.py [hunk 1]
```diff
@@ -64,6 +65,65 @@ def yank_path(self, path):
                 if path in key[0]:
                     del self._caches[func][key]
 
+class ByteRangeReader(object):
+
```
Comment (percyliang) [inline codalab/lib/bundle_fuse.py]: Add high-level description.

File: codalab/lib/bundle_fuse.py [hunk 2]
```diff
@@ -64,6 +65,65 @@ def yank_path(self, path):
...
+                print 'expired'
+                self.cache.pop(key)
+
+        # grab from client
+        byte_range = (chunk_id * self.chunk_size, chunk_id * self.chunk_size + self.chunk_size - 1)
+        with closing(self.client.fetch_contents_blob(self.bundle_uuid, path, byte_range)) as contents:
+            arr = contents.read()
+
+        if len(arr) == self.chunk_size: # only cache if fetched full chunk
+            print 'put to cache'
```
Comment (percyliang) [inline codalab/lib/bundle_fuse.py]: Probably don't want these print statements, or guard them by some verbose flag.

**Review by percyliang — APPROVED** [review]
Looks generally reasonable.  In general, more documentation, especially about the performance characteristics about how the code is exercised in practice, would be nice.

Question: if you do a cp of a 30 MB file, what is `length` usually?  If this is larger than 1 MB, then we could have one call that grabs multiple chunks.  It seems like the REST API is still a bottleneck given the difference between using FUSE and `cl down`.  Probably worth quantifying this.

**Comment by fabeschan** [comment]
Thanks @percyliang , I'll do a cleanup and add documentation. 

`length` is usually around ~13000 bytes

File: codalab/lib/bundle_fuse.py [hunk 3]
```diff
@@ -64,6 +65,65 @@ def yank_path(self, path):
...
+        arr = ''
+        for chunk_id in range(start_chunk, end_chunk + 1):
+            arr += self._fetch_chunk(path, chunk_id)
+        return arr[offset - start_chunk * self.chunk_size:offset - start_chunk * self.chunk_size + length]
+
+    def _fetch_chunk(self, path, chunk_id):
+        '''
+        Fetch a chunk from the cache, or with the client as necessary.
+        Refreshes chunks that are older than self.timeout
+        Only save full chunks to the cache. Partial chunks are not cached.
```
Comment (percyliang) [inline codalab/lib/bundle_fuse.py]: Say that partial chunks are only at the end of the file because they could grow...

Make the assumptions clear.

Comment (fabeschan) [reply codalab/lib/bundle_fuse.py]: done

File: codalab/lib/bundle_fuse.py (hunk 1 above)
Comment (fabeschan) [reply codalab/lib/bundle_fuse.py]: done

**Comment by maxwang7** [comment]
LGTM! I didn't do any perf testing, but functionally it works for me on Ubuntu 14.04.

**Pull request merged**